### ISSUE TITLE
dart2js: After "lhs ??= rhs", lhs is null even though rhs is not null.

### ISSUE THREAD

**Issue by johnchen902** [dart-lang/sdk#47166]
The following program should definitely prints `2`, but it prints `null` under dart2js. For one thing, the return type of `triggerTheProblem` is `int`, not `int?`!

```dart
void maybeThrow(bool doThrow) {
  if (doThrow) throw Exception('omg!');
}

int triggerTheProblem(bool doThrow) {
  int? x = 1;
  try {
    x = null;
    maybeThrow(doThrow);
    x = 1;
  } catch (e) {
    // ignore
  }
  x ??= 2;
  (() => x)();
  return x;
}

void main() {
  print(triggerTheProblem(true));
}
```

This can be reproduced on DartPad ("Based on Flutter 2.5.0 Dart SDK 2.14.0" at the bottom-right corner), and I'm using Linux and Chrome 93.0.4577.63.

This is a reduced sample, so I'm relatively sure this is actually a dart2js bug (not DartPad's or V8's), and it also affects whatever Dart version my company is using (2.15.0 if I'm not mistaken).

**Comment by lrhn** [comment]
Seems like an overly optimistic assumption that `x` is definitely not null after the `try`/`catch`, so the `??= 2` can be ignored.

That assumption is not safe. A `catch` which doesn't itself throw means that you cannot assume anything about how much of the `try` block has been executed.
It's not enough to assume that it has either been executed entirely or not. Every statement of a try block has a potential control flow to the `catch`, and when the catch doesn't throw, also into the following code.

**Comment by johnchen902** [comment]
> It seems like the assignment of null to x is somehow delayed, because the value of x is 1 when leaving the try/catch.

On the contrary, my educated guess is that SsaLoadElimination sometimes replaces the load of `x` with `1`, even though the actual value of `x` is `null`.

**Comment by rakudrama** [comment]
Good call, @johnchen902 , disabling SsaLoadElimination makes the program work correctly.
I'll look into this further next week.

**Comment by rakudrama** [comment]
Fixed via ff617b9505a4f3f44181850b58920cbd658e97d4
